staging_service/kb_auth_client.py

```diff
@@ -0,0 +1,159 @@
+"""
+A client for the KBase Auth2 server.
+"""
+
+# Mostly copied from https://github.com/kbase/cdm-task-service with a few tweaks,
+# - included the require_string function
+# - removed roles check and management
+# TODO make a KBase auth library
+
+import aiohttp
+from cacheout.lru import LRUCache
+import logging
+import time
+from typing import Self
+
+
+async def _get(url: str, headers: dict[str, str]):
+    # TODO PERF keep a single session and add a close method
+    async with aiohttp.ClientSession() as session:
+        async with session.get(url, headers=headers) as r:
+            await _check_error(r)
+            return await r.json()
+
+
+async def _check_error(r):
+    if r.status != 200:
+        try:
+            j = await r.json()
+        except Exception:
+            err = "Non-JSON response from KBase auth server, status code: " + str(r.status)
+            logging.getLogger(__name__).info("%s, response:\n%s", err, r.text)
+            raise IOError(err)
+        # assume that if we get json then at least this is the auth server and we can
+        # rely on the error structure.
+        err = j["error"].get("appcode")
+        if err == 10020:  # Invalid token
+            raise InvalidTokenError("KBase auth server reported token is invalid.")
```

Comment:
It looks like everywhere else is raising webserver errors (which I'm not a fan of, but anyway...). Is that going to cause a problem here?

Reply:
I'm a little annoyed that the existing test suite didn't catch that. Again, this PR is big enough. I'll make a fresh PR with some app integration tests.

Reply:
Fixed.

Reply:
I'm not grokking your error mapping

ValueError -> 403
 - This is a programming error, a user can never cause this. It should never happen but it it does should be a 5XX

InvalidTokenError should map to a 401

You don't map IOError or InvalidUserError but maybe that's intentional

Reply:
Unless you're using the _require_string fn to look for missing parameters. If so I wouldn't do that and put in an explicit check upstream for the username / token / whatever that throws a 400. The _require_string fn is just a programming safety check, not an API inputs check

Reply:
If you want the auth class to be responsible for checking valid inputs from user land, then the class should throw specific errors for the problem - ValueError is too generic and could be thrown for things that aren't a missing token or missing username. I'd probably add MissingTokenError and MissingUsernameError